sympy/physics/mechanics/tests/test_joint.py

```diff
@@ -0,0 +1,228 @@
+    assert C.masscenter.pos_from(rod.masscenter) == lA*A.z
+    assert rod.masscenter.pos_from(plate.masscenter) == - lC * A.z
+
+    #Check Linear Velocities
+    assert rod.masscenter.vel(N) == (h/4 - lB/2)*omega*A.x
+    assert plate.masscenter.vel(N) == ((h/4 - lB/2)*omega + (h/4 + lB/2)*omega)*A.x
+
+def test_pinjoint_arbitrary_axis():
+    def generate_body():
+        N = ReferenceFrame('N')
+        A = ReferenceFrame('A')
+        P = Body('P', frame=N)
+        C = Body('C', frame=A)
+        return N, A, P, C
+
+    theta, omega = dynamicsymbols('theta_J, omega_J', real=True)
+
+    #When the bodies are attached though masscenters but axess are opposite.
+    N, A, P, C = generate_body()
+    PinJoint('J', P, C, child_axis=-A.x)
+
+    assert -A.x.angle_between(N.x) == 0
+    assert -A.x.express(N) == -N.x
+    assert A.dcm(N) == Matrix([[1, 0, 0], [0, cos(theta), sin(theta)], [0, -sin(theta), cos(theta)]])
+    assert A.ang_vel_in(N) == omega*N.x
+    assert A.ang_vel_in(N).magnitude() == Abs(omega)
+    assert C.masscenter.pos_from(P.masscenter) == 0
+    assert C.masscenter.pos_from(P.masscenter).express(N).simplify() == 0
+    assert C.masscenter.vel(N) == 0
+
+    #When axes are different and parent joint is at masscenter but child joint is at a unit vector from
+    #child masscenter.
+    N, A, P, C = generate_body()
+    PinJoint('J', P, C, child_axis=A.y, child_joint_pos=A.x)
+
+    assert A.y.angle_between(N.x) == 0 #Axis are aligned
+    assert A.y.express(N) == N.x
+    assert A.dcm(N) == Matrix([[0, -cos(theta), -sin(theta)], [1, 0, 0], [0, -sin(theta), cos(theta)]])
+    assert A.ang_vel_in(N) == omega*N.x
+    assert A.ang_vel_in(N).express(A) == omega * A.y
+    assert A.ang_vel_in(N).magnitude() == Abs(omega)
+    assert expand_mul(A.ang_vel_in(N).angle_between(A.y)) == acos(omega/Abs(omega))
+    assert C.masscenter.vel(N) == omega*A.z
+    assert C.masscenter.pos_from(P.masscenter) == -A.x
+    assert C.masscenter.pos_from(P.masscenter).express(N).simplify() == cos(theta)*N.y + sin(theta)*N.z
+    assert C.masscenter.vel(N).angle_between(A.x) == pi/2
+
+    #Similar to previous case but wrt parent body
+    N, A, P, C = generate_body()
+    PinJoint('J', P, C, parent_axis=N.y, parent_joint_pos=N.x)
+
+    assert N.y.angle_between(A.x) == 0 #Axis are aligned
+    assert N.y.express(A) ==  A.x
+    assert A.dcm(N) == Matrix([[0, 1, 0], [-cos(theta), 0, sin(theta)], [sin(theta), 0, cos(theta)]])
+    assert A.ang_vel_in(N) == omega*N.y
+    assert A.ang_vel_in(N).express(A) == omega* A.x
+    assert A.ang_vel_in(N).magnitude() == Abs(omega)
+    assert expand_mul(A.ang_vel_in(N).angle_between(A.x)) == acos(omega/Abs(omega))
+    assert C.masscenter.vel(N).simplify() == - omega*N.z
+    assert C.masscenter.pos_from(P.masscenter) == N.x
+
+    #Both joint pos id defined but different axes
+    N, A, P, C = generate_body()
+    PinJoint('J', P, C, parent_joint_pos=N.x, child_joint_pos=A.x, child_axis=A.x+A.y)
```

Comment:
Looking only at the orientation for this example with the child rotation axis as `rC = A.x + A.y` I have this figure:
![image](https://user-images.githubusercontent.com/276007/123761961-99c4d200-d8c2-11eb-891a-b77459fd6833.png)

`uC = cross(As.x, N.x).normalize()`

My code to generation the orientation and angular velocity expressions is:

```
In [1]: import sympy as sm

In [2]: import sympy.physics.mechanics as me

In [3]: N = me.ReferenceFrame('N')

In [4]: A = me.ReferenceFrame('A')

In [5]: As = me.ReferenceFrame('As')

In [6]: rC = A.x + A.y

In [7]: As.orient_axis(N, -N.z, sm.pi/4)  # note the negative N.z here

In [8]: theta = me.dynamicsymbols('theta')

In [9]: A.orient_axis(As, N.x, theta)

In [11]: sm.simplify(A.dcm(N))
Out[11]: 
Matrix([
[sqrt(2)/2, -sqrt(2)*cos(theta(t))/2, -sqrt(2)*sin(theta(t))/2],
[sqrt(2)/2,  sqrt(2)*cos(theta(t))/2,  sqrt(2)*sin(theta(t))/2],
[        0,           -sin(theta(t)),            cos(theta(t))]])

In [13]: A.ang_vel_in(N).express(N)
Out[13]: Derivative(theta(t), t)*N.x

In [14]: A.ang_vel_in(N).express(A).simplify()
Out[14]: sqrt(2)*Derivative(theta(t), t)/2*A.x + sqrt(2)*Derivative(theta(t), t)/2*A.y

In [16]: A.x.express(N).simplify()
Out[16]: sqrt(2)/2*N.x - sqrt(2)*cos(theta(t))/2*N.y - sqrt(2)*sin(theta(t))/2*N.z

In [16]: A.x.express(N).simplify()
Out[16]: sqrt(2)/2*N.x - sqrt(2)*cos(theta(t))/2*N.y - sqrt(2)*sin(theta(t))/2*N.z

In [17]: A.z.express(N).simplify()
Out[17]: - sin(theta(t))*N.y + cos(theta(t))*N.z

In [18]: A.y.express(N).simplify()
Out[18]: sqrt(2)/2*N.x + sqrt(2)*cos(theta(t))/2*N.y + sqrt(2)*sin(theta(t))/2*N.
```